### ISSUE TITLE
How to best debug "test failed:" type errors?

### ISSUE THREAD

**Issue by utcke** [Dobiasd/frugally-deep#88]
Hi,

is there any reasonable way to debug "test failed: output=0 pos=0,0,0 value=-1041.18 target=-401.894" type errors?  

Disabeling the tests does indeed result in the wrong output, so how about going to debug this?  

The network structure ist quite small (just a Conv2D and 5 SeperableConv2D, all with BatchNormalization and activation relu, and all cobbled together using skip connections, and a Dense Layer, learning a 16-dimensional embedding for a 64x64 image), about 400K for the json file.  I'm a bit baffled why this should fail.  

Any hints where to start looking?

I would really love to get this working, as fdeep seems the perfect tool for what I want.

Thanks a lot

Sven

**Comment by Dobiasd** [comment]
Hi Sven,

have you tried with

```cpp
#define FDEEP_FLOAT_TYPE double
#include <fdeep/fdeep.hpp>
```
?

**Comment by Dobiasd** [comment]
OK, so I guess it did not help. :)

What is the range of the numbers in the input tensors, the model is trained with in Keras?

**Comment by utcke** [comment]
The range is 0-1, the model was trained with Keras.  It it helps, this is TF 1.10 and keras 2.2.4.  The gcc is 5.4.0

Some more information which I think is probably completely irrelevant: The model originally had a lambda layer doing normalisation at the end.  I stripped that layer out before writing the h5, and both a run of that model before writing as well as after reading in in convert_model.py produces the expected result (except for the now missing normalization, obviously).

Any more info I could give you?  I'll try the minimal example from your README next...

**Comment by utcke** [comment]
Ok, I ran the minimal example given in your README, compiling using

> g++ -I/home/sutcke/include -I/home/sutcke/include/eigen3 -std=c++14 -O3 main.cpp -L/usr/X11R6/lib -lm -lpthread -lX11 -lstdc++fs -o main

And the output was:
> Loading json ... done. elapsed time: 0.000096 s
> Running test 1 of 1 ... done. elapsed time: 0.000026 s
> Loading, constructing, testing of fdeep_model.json took 0.000224 s overall.
> [[[[0.8499]], [[0.1093]], [[0.0408]]]]

Which I suppose means it worked alright.  So I'm a bit baffled...

**Comment by utcke** [comment]
Any ideas (sort of anxiously waiting :-)

**Comment by Dobiasd** [comment]
Sorry for the delay. I'm currently offline(ish) for two weeks. After that I can happily debug into the issue. Hope this is not too late for you.

One attempt you could already do is try to use a different input number range, i.e. -1 to 1 instead of 0 to 1, for training and prediction.

**Comment by utcke** [comment]
Sorry to be such a bother.  It is rather late, but that’s the price for free (in fact I’m mighty glad you’re willing to help at all).  I’ll keep you informed if I can find out anything…

Mit freundlichen Grüßen / Best regards

Dr.-Ing. Sven Utcke
DERMALOG Identification Systems GmbH
Senior Systems Engineer

Mittelweg 120 | 20148 Hamburg
Fon +49 40 413227 0

Fax +49-40-413227-40

Mail sven.utcke@dermalog.com<mailto:sven.utcke@dermalog.com>

Web www.dermalog.com<https://www.dermalog.com>


This e-mail may contain information that is confidential or proprietary to DERMALOG Identification Systems GmbH. Any unauthorized disclosure, distribution or other use is prohibited.

P Please consider the environment before printing this email

**Comment by Dobiasd** [comment]
OK, so you already tested
- `double` instead instead of `float`
- skipping the tests and trying with real input values

Both things did not help.

My guess is it has to do with some numerical effects of multiplication/division with very large/small values. So the different input value range might really help.

If this does not help you can additionally check if your model has some unrealistic large weights somewhere. At least the desired output in the test (`-401.894`) suggests an investigation into this direction.

If yoy can quickly re-train the model, you could try to use elu instead of relu, or even use a CPU if you used a GPU before and vice versa.

**Comment by utcke** [comment]
Hmm.
I gave the problem to a colleague, and he found that with all models he tried he could get things to run by setting 

const auto model = fdeep::load_model("file.json", true, NULL, 0.01);

so I now tried to replicate the problem again on my original model and found that I could not duplicate the problem (which presumably means I overwrote my orginal model somewhere along the line, even though I can not remember doing so).
So I'm tempted to assume that this was a freak model, fdeep works in general on our type models, and I'm not sure how much sense it makes to keep this bug open (although it would be interesting to find out why the 0.01 is needed, and how to improve on that).

Sorry for the noise...

**Comment by Dobiasd** [comment]
No problem. I'm happy you were able to make it work. And I hope your further experience with fdeep will be more pleasing. :)

Since your model seems to output values a few orders of magnitude larger than the usual 0.0 to 1.0 range, needing a larger epsilon for the tests seems at least somewhat plausible. Is the larger epsilon also needed when using
```cpp
#define FDEEP_FLOAT_TYPE double
#include <fdeep/fdeep.hpp>
```
?

Also, if you can upload a version of your model to reproduce the effect, I could try to find out something.

**Comment by utcke** [comment]
That epsilon of 0.01 was with double (but going to float gives about the same results).

The "real" model does output values between -1 and 1, but the last layer was a lambda layer doing normalization, so I had to remove that one for fdeep, hence the bigger values.

If I ever find the non-working model again I will be more than happy to upload it if you are willing to take the time to look into it.  Thanks!

**Comment by Dobiasd** [comment]
OK, just re-open the issue then. :+1:
